### PR TITLE
ipc: optimize memory usage for IPC notifications

### PR DESCRIPTION
**Pull request by tlauda**
Refactors IPC notification sending in order to minimize allocated
memory space. Instead of statically allocating the whole queue of
twelve ipc_msg structures, let's do it dynamically if there is the
actual need of sending message to host.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

File: src/ipc/handler.c [hunk 1]
```diff
@@ -1250,31 +1228,43 @@ static inline struct ipc_msg *msg_find(struct ipc *ipc, uint32_t header,
...
+		}
 
-	if (msg == NULL) {
-		trace_ipc_error("ipc: msg hdr for 0x%08x not found "
-				"replace %d", header, replace);
-		ret = -EBUSY;
-		goto out;
+		/* create message */
+		msg = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			      SOF_MEM_CAPS_RAM, sizeof(*msg));
```
Comment (paulstelian97) [inline src/ipc/handler.c]: @dbaluta I think allocating memory once per IPC may be a concern since that's quite a few allocations, right? What do you think?

File: src/drivers/intel/baytrail/ipc.c [hunk 2]
```diff
@@ -125,39 +125,37 @@ static void ipc_platform_complete_cmd(void *data)
 	}
 }
 
-void ipc_platform_send_msg(struct ipc *ipc)
+void ipc_platform_send_msg(void)
```
Comment (paulstelian97) [inline src/drivers/intel/baytrail/ipc.c]: I'm not sure removing the ipc parameter and removing the `dsp_pending` flag are related changes that should remain in the same commit?

File: src/include/sof/drivers/ipc.h [hunk 3]
```diff
@@ -143,7 +143,7 @@ int ipc_stream_send_xrun(struct comp_dev *cdev,
 	struct sof_ipc_stream_posn *posn);
 
 int ipc_queue_host_message(struct ipc *ipc, uint32_t header, void *tx_data,
-			   size_t tx_bytes, uint32_t replace);
+			   size_t tx_bytes, bool replace);
```
Comment (paulstelian97) [inline src/include/sof/drivers/ipc.h]: Change from uint32_t to bool is a refactor, I'd prefer it in a separate commit dedicated to all such refactors.

File: src/ipc/handler.c (hunk 1 above)
Comment (tlauda) [reply src/ipc/handler.c]: 1. Stream positions are not sent every 1 ms, but rather every period specified by host (I've noticed that 40 ms is popular value).
2. Position updates are also written into memory window, so on the platforms with such capability we should really stop spamming IPCs.
3. The queue is limited to 12 messages anyway, so right now you also can lose the notification if the host won't process it fast enough or DSP has no idle cycles to go through queue. Nothing changes in this department and I would even go as far as saying that having full queue is more probable than not having free space on runtime heap.

Comment (paulstelian97) [reply src/ipc/handler.c]: Yeah the issue is if there is some weird issue with the modules/components (future work) we may overallocate runtime memory and end up not having enough memory to be able to send 12 messages. I mean I'm not blocking over it but there are clients which may complain about this.

Allocating at prepare time may be acceptable, but playback stopping due to some leak from a component which isn't even upstreamed means reliability takes a (small) dip.

Your point about the memory window, yeah I think I should figure out how to make a memory window myself. Is there some section of the mailbox itself that I can reuse or I should update the memory map (taking advantage of the huge shared memory on i.MX) and just add a new region specifically for this? How exactly does it work?

"you can also lose the notification if [...] DSP has no idle cycles to go through queue" That is easier to hit during testing though -- in fact hitting it made me enable the caches when it had to happen. Still, losing the notification will cause serious problems of quality. The allocation issues may in certain situations which third party code be hit only under certain circumstances.

Comment (dbaluta) [reply src/ipc/handler.c]: @tlauda @paulstelian97 I think Paul has a point in this but because the queue size is pretty small I don't think it matters. If allocations will start failing at runtime we are in trouble. Another, point to consider is that allocations will take time. 

@paulstelian97 I do think it will worth considering by using the shared memory to post position updates instead of sending them through IPCs.

Comment (tlauda) [reply src/ipc/handler.c]: @paulstelian97 @dbaluta Changed it back to for now. We can revisit this in the future.

File: src/drivers/imx/ipc.c [hunk 4]
```diff
@@ -125,16 +125,16 @@ static void ipc_platform_complete_cmd(void *data)
 	}
 }
 
-void ipc_platform_send_msg(struct ipc *ipc)
+void ipc_platform_send_msg(void)
```
Comment (kv2019i) [inline src/drivers/imx/ipc.c]: Could you enhance the commit message a bit and explain why you are removing the argument. It seems we have a static _ipc instance already defined, but why we are changing now, that would be good to explain in the commit message.

File: src/drivers/intel/cavs/ipc.c [hunk 5]
```diff
@@ -124,7 +124,7 @@ static void ipc_irq_handler(void *arg)
 			  ipc_read(IPC_DIPCCTL) | IPC_DIPCCTL_IPCIDIE);
 
 		/* send next message to host */
-		ipc_process_msg_queue();
+		ipc_platform_send_msg();
```
Comment (kv2019i) [inline src/drivers/intel/cavs/ipc.c]: This seems like an additional change that does not belong to this commit.

File: src/schedule/task.c [hunk 6]
```diff
@@ -47,7 +47,7 @@ enum task_state task_main_master_core(void *data)
 		wait_for_interrupt(0);
 
 		/* now process any IPC messages to host */
-		ipc_process_msg_queue();
+		ipc_platform_send_msg();
```
Comment (kv2019i) [inline src/schedule/task.c]: Same as above, why this change here?

File: src/drivers/imx/ipc.c [hunk 7]
```diff
@@ -29,36 +29,6 @@ struct ipc_data {
 	struct ipc_data_host_buffer dh_buffer;
 };
```
Comment (kv2019i) [inline src/drivers/imx/ipc.c]: Should this be done in same commit as patch that removes the "cb" callback. I think bisect won't work otherwise.

Typo in commit: "Clean up IPC drivers by removing..."

File: src/include/sof/drivers/ipc.h [hunk 8]
```diff
@@ -17,6 +17,7 @@
 #include <sof/trace/trace.h>
 #include <ipc/header.h>
 #include <user/trace.h>
+#include <stdbool.h>
```
Comment (kv2019i) [inline src/include/sof/drivers/ipc.h]: Language in commit message: "Changes replace argument" -> "Replace argument of..."

File: src/ipc/handler.c [hunk 9]
```diff
@@ -1245,8 +1245,9 @@ int ipc_queue_host_message(struct ipc *ipc, uint32_t header, void *tx_data,
 			   size_t tx_bytes, bool replace)
 {
 	struct ipc_msg *msg = NULL;
-	uint32_t flags, found = 0;
+	bool found = false;
 	int ret = 0;
+	uint32_t flags;
```
Comment (kv2019i) [inline src/ipc/handler.c]: Reverse xmas-tree declaration order, flags should be ret.

File: src/ipc/handler.c [hunk 10]
```diff
@@ -1245,8 +1245,9 @@ int ipc_queue_host_message(struct ipc *ipc, uint32_t header, void *tx_data,
 			   size_t tx_bytes, bool replace)
 {
 	struct ipc_msg *msg = NULL;
-	uint32_t flags, found = 0;
+	bool found = false;
```
Comment (kv2019i) [inline src/ipc/handler.c]: "bool replaced = false" perhaps? Ok with found as well, just a bit confusing to read the function body.

File: src/ipc/ipc.c [hunk 11]
```diff
@@ -471,9 +472,12 @@ int ipc_init(struct sof *sof)
 	list_init(&sof->ipc->shared_ctx->msg_list);
 	list_init(&sof->ipc->shared_ctx->comp_list);
 
-	for (i = 0; i < MSG_QUEUE_SIZE; i++)
-		list_item_prepend(&sof->ipc->shared_ctx->message[i].list,
+	for (i = 0; i < MSG_QUEUE_SIZE; i++) {
```
Comment (kv2019i) [inline src/ipc/ipc.c]: @tlauda What's the benefit if we anyways allocate MSG_QUEUE_SIZE of messages here? Earlier it's in rzalloc() of shared_ctx, now here...

File: src/drivers/imx/ipc.c (hunk 7 above)
Comment (tlauda) [reply src/drivers/imx/ipc.c]: It's different callback.

File: src/include/sof/drivers/ipc.h (hunk 8 above)
Comment (tlauda) [reply src/include/sof/drivers/ipc.h]: Done.

File: src/drivers/imx/ipc.c (hunk 4 above)
Comment (tlauda) [reply src/drivers/imx/ipc.c]: Done.

File: src/schedule/task.c (hunk 6 above)
Comment (tlauda) [reply src/schedule/task.c]: Done.

File: src/ipc/handler.c (hunk 9 above)
Comment (tlauda) [reply src/ipc/handler.c]: Done.

File: src/ipc/ipc.c (hunk 11 above)
Comment (tlauda) [reply src/ipc/ipc.c]: @kv2019i One of the benefits is getting them cache line aligned for free, but mostly it's a preparation for further multicore based changes. I'm planning on adding separate heap for shared memory, providing separate sections for shared static data and also extracting "uncached" related code to specific platforms.

File: src/drivers/imx/ipc.c (hunk 7 above)
Comment (kv2019i) [reply src/drivers/imx/ipc.c]: Ack @tlauda

File: src/ipc/ipc.c (hunk 11 above)
Comment (paulstelian97) [reply src/ipc/ipc.c]: @tlauda About the uncached... I'm thinking about the fact that the imx platform doesn't have uncached memory (save for the actual IO-mapped devices) so the platform specific macros are lying on whether the memory is cached or not (not sure what lie they tell right now)

Comment (kv2019i) [reply src/ipc/ipc.c]: @tlauda Ack ok. I got misled a bit with the PR title "optimize memory usage", but the commit message for this patch is actually more accurate. Others seem to agree, so good for me as well.

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
CI usual suspects
